sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -74,10 +84,15 @@ case class Generate(
           } else {
             outputRows.map(or => joinProjection(joinedRow(row, or)))
           }
+        } ++ LazyIterator(() => boundGenerator.terminate()).map { row =>
+          joinedRow(nullOriginalRow, row)
         }
       }
     } else {
-      child.execute().mapPartitions(iter => iter.flatMap(row => boundGenerator.eval(row)))
+      child.execute().mapPartitions(iter =>
+        iter.flatMap(row => boundGenerator.eval(row)) ++
+          LazyIterator(() => boundGenerator.terminate())
```

Comment:
@maropu thanks for the suggestion, I'd like to try that locally. BTW can you confirm if this PR work for you, as we talked offline? Just to double confirm the concern of @viirya.
Thanks.